File: csv-combiner.test.js
const fs = require('fs')
const path = require('path')
const Papa = require('papaparse')
const { handleCSVFiles } = require('./csv-combiner.js')

const outputFile = 'combined_test.csv'

function wait(ms) {
  return new Promise(resolve => setTimeout(resolve, ms))
}

const TestFileCreation = async () => {
  /*
    Purpose: Runs primary function in csv-combiner.js with test data
    and checks to make sure new file is created

    Output: true if file exists / false if file cannot be found
    */

  const inputFiles = fs
    .readdirSync(path.resolve(__dirname, 'fixtures_test'))
    .map(file => path.resolve(__dirname, 'fixtures_test', file))
  handleCSVFiles(outputFile, inputFiles)
  await wait(100)
  try {
    if (fs.existsSync(outputFile)) return true
  } catch (err) {
    return false
  }
}

const TestFileRowCount = async () => {
  /*
    Purpose: Used to compare length of new file and the 
    combined length of the files it was merged from

    Output: returns the length of newly created file
    */

  await wait(100)
  let data = fs.readFileSync(outputFile)
  data = data.toString()
  data = Papa.parse(data).data
  return data.length
}

const InputFileCount = () => {
  /*
    Purpose: Used to compare length of new file and the 
    combined length of the files it was merged from

    Output: returns combined length of original
    */
  const inputFiles = fs
    .readdirSync(path.resolve(__dirname, 'fixtures_test'))
    .map(file => path.resolve(__dirname, 'fixtures_test', file))

  let result = 0
  for (let i = 0; i < inputFiles.length; i++) {
    let data = fs.readFileSync(inputFiles[i])
    data = data.toString()
    data = Papa.parse(data).data
    result = result + data.length
  }
  //Backout extra headers minus the one header we used
  return result - inputFiles.length - 1
}

test('#1 Test that new file is created', () => {
  return expect(TestFileCreation()).resolves.toBe(true)
})

test('#2 Test that files has correct amount of rows', () => {
  return expect(TestFileRowCount()).resolves.toBe(InputFileCount())
})

File: csv-combiner_test.go
package main

import (
	"encoding/csv"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"
	"strconv"
	"testing"
)

func TestFileRowCount(t *testing.T) {
	/*
	   Purpose: Used to compare length of new file and the
	   combined length of the files it was merged from

	   Output: PASS if combined file and individul files are the same length else FAIL
	*/
	path, err := os.Getwd()
	if err != nil {
		log.Println(err)
	}
	files, err := ioutil.ReadDir(path + "/fixtures_test")
	if err != nil {
		log.Println(err)
	}
	inputFiles := make([]string, 0)
	for _, f := range files {
		inputFiles = append(inputFiles, "//fixtures_test/"+f.Name())
	}

	outputFile := "combined_test.csv"

	handleCSVFiles(outputFile, inputFiles)

	inputCount := 0
	for i := range inputFiles {
		f, err := os.Open(path + inputFiles[i][1:])
		if err != nil {
			log.Fatal(err)
		}
		defer f.Close()
		r := csv.NewReader(f)
		for {
			_, err := r.Read()
			if err == io.EOF {
				break
			}
			inputCount++
		}
	}
	//We will have one header for the final output, remove extra headers from inputCount
	inputCount = inputCount - len(inputFiles) + 1
	f, err := os.Open(path + "/" + outputFile)
	if err != nil {
		log.Fatal(err)
	}
	defer f.Close()
	r := csv.NewReader(f)
	outputCount := 0
	for {
		_, err := r.Read()
		if err == io.EOF {
			break
		}
		outputCount++
	}
	if inputCount != outputCount {
		fmt.Println("INPUT FILE LENGTH: " + strconv.Itoa(inputCount))
		fmt.Println("OUTPUT FILE LENGTH: " + strconv.Itoa(outputCount))
		t.Fail()
	}

}
